compositor_pipeline/src/pipeline/input/mp4.rs

```diff
@@ -89,72 +87,312 @@ impl Mp4 {
+                            .name("mp4 reader - audio".to_string())
+                            .spawn(move || {
+                                let _span = span.enter();
+                                let mut track = match track {
+                                    TrackProvider::Value(track) => track,
+                                    TrackProvider::Handle(handle) => handle.join().unwrap(),
+                                };
+                                for (mut chunk, duration) in track.chunks() {
+                                    chunk.pts += offset;
+                                    chunk.dts = chunk.dts.map(|dts| dts + offset);
+                                    last_sample_pts.fetch_max(
+                                        (chunk.pts + duration).as_nanos() as u64,
+                                        Ordering::Relaxed,
+                                    );
+                                    trace!(pts=?chunk.pts, "MP4 reader produced an audio chunk.");
+                                    if sender.send(PipelineEvent::Data(chunk)).is_err() {
+                                        debug!("Failed to send a audio chunk. Channel closed.")
+                                    }
+                                    if should_close.load(Ordering::Relaxed)
+                                        || should_close_input.load(Ordering::Relaxed)
+                                    {
+                                        break;
+                                    }
+                                    // TODO: send flush
+                                }
+                                let _ = finished_track_sender.send(TrackType::Audio);
+                                track
+                            })
+                            .unwrap()
+                    });
+
+                match finished_track_receiver.recv().unwrap() {
+                    TrackType::Video => {
+                        video_track =
+                            Some(TrackProvider::Value(video_thread.unwrap().join().unwrap()));
+                        should_close.store(true, Ordering::Relaxed);
+                        if let Some(audio_thread) = audio_thread {
+                            audio_track = Some(TrackProvider::Handle(audio_thread));
+                        }
+                    }
+                    TrackType::Audio => {
+                        audio_track =
+                            Some(TrackProvider::Value(audio_thread.unwrap().join().unwrap()));
+                        should_close.store(true, Ordering::Relaxed);
+                        if let Some(video_thread) = video_thread {
+                            video_track = Some(TrackProvider::Handle(video_thread));
+                        }
+                    }
+                }
+                if has_audio {
+                    offset = Duration::from_nanos(last_audio_sample_pts.load(Ordering::Relaxed));
```

Comment:
If an audio track is present use it specifically. Using lasdt video sample or last sample in general is leading to audio artifacts on the start of a new track

Reply:
buffers between queue<->decoder and decoder-input are in chunks, if chunks (or frame/sample batches) do not represent similar duration then timestamps for audio and video here can differ significantly. 

For mp4 (15 fps) I tested, audio processing was behind the video by over 2 seconds, because the buffer can fit a lot more video at low fps.

2 address that we would have to use channels between elements that have sized based on media duration instead.

Reply:
oh, that makes a lot of sense.